src/boot/zlib.c

```diff
@@ -0,0 +1,1290 @@
+        unsigned int j;
+
+        /* Fill in any symbols of this code length. */
+        for (j = 0; j < symbols; j++) {
+            if (lengths[j] == i) {
+                table[index++] = j;
+            }
+        }
+
+        /* Fill in remaining (internal) nodes for this length. */
+        for (j = next_code; j < code_limit; j++) {
+            table[index++] = ~next_index;
+            next_index += 2;
+        }
+    }  /* for each code length */
+
+    /* Return success. */
+    return 1;
+}
+
+/*************************************************************************/
+/*************************************************************************/
+
+/* End of tinflate.c */
+
+struct decoder
+{
+	unsigned char   buf[1024*KIB];   /* intermediate buffer for loading  */
+	unsigned char  *buf_end;     /* pointer that exists for the sole *
+	                              * purpose of getting size of `buf` */
+	unsigned        pstart;      /* offset of next read from rom     */
+	unsigned int    remaining;   /* remaining size of file           */
+	unsigned char  *buf_limit;   /* points to end of scannable area  *
+	                              * of buf; this prevents yaz parser *
+	                              * from overflowing                 */
+#if MAJORA
+	unsigned char  *dst_end;     /* end of decompressed block        */
+#endif
+};
+
+static struct decoder dec;
+
+/* request more compressed data */
+static inline unsigned refill(void)
+{
+	unsigned int    size;
+	unsigned char  *dst = dec.buf;
+	
+	/* calculate size for next read */
+	size = sizeof(dec.buf);
+	
+	/* if it exceeds remaining file size, use that */
+	if (dec.remaining < size)
+		size = dec.remaining;
+	
+	/* read file from rom */
+	DmaMgr_DmaRomToRam(dec.pstart, dst, size);
+	
+	dec.pstart += size;
+	dec.remaining -= size;
+	
+	return size;
+}
+
+/* main driver */
+//void ENTRY(void *src_, void *dst_, unsigned sz)
+void Zlib_Decompress(uintptr_t romStart, u8* dst, size_t sz)
+{
+	unsigned src = romStart;
+	static DecompressionState state = {-1};
+	
+	/* skip header */
+	src += 8;
+	sz -= 8;
+	
+	/* initialize decoder structure */
+	dec.buf_end = dec.buf + sizeof(dec.buf);
+	dec.pstart = src;
+	dec.remaining = sz;
+
+	/* clear decompression state buffer */
+	state.state	    = INITIAL;
+	state.out_ofs   = 0;
+#ifdef WANT_CRC
+	state.crc	    = 0;
+#endif
+	state.bit_accum = 0;
+	state.num_bits  = 0;
+	state.final	    = 0;
+	/* no other fields need to be cleared */
+	
+	while (1)
+	{
+		int dstMax = 1024 * 1024 * 32; /* max size of any file: 32mb */
+		unsigned long size = 0; /* XXX must be zero-initialized */
+		unsigned long crc_ret;
+		unsigned readSize;
+		int result;
+		readSize = refill();
+		result = tinflate_partial(
+			dec.buf, readSize,
+			dst, dstMax,
+			&size, &crc_ret,
+			&state, sizeof(state)
+		);
+		dst += size;
+		if (!result)
+			break;
+	}
+	
+#if MAJORA
+	dec.buf_end = 0;
+#endif
```

Comment:
since you removed MM stuff from other algorithms you should probably remove it here too for consistency